src/deepsparse/v2/pipeline.py

```diff
@@ -13,88 +13,113 @@
 # limitations under the License.
 
 
-from typing import List
-
-from pydantic import BaseModel, Field, PrivateAttr
+from typing import Any, Dict, List, Optional, Union
 
 from deepsparse.v2.operators import Operator
 from deepsparse.v2.routers import Router
 from deepsparse.v2.schedulers import OperatorScheduler, SchedulerGroup
+from deepsparse.v2.utils import Context
 
 
 __all__ = ["Pipeline"]
 
 
-class Pipeline(BaseModel):
-    """
-    Pipeline accepts a series of operators, schedulers, and a router which define
-    an end to end ML transformation.
-
-    Calling a pipeline runs these transformations
+class Pipeline(Operator):
     """
+    Pipeline accepts a series of operators, schedulers, and a router. Calling a pipeline
+    will use the router to run through all the defined operators.
 
-    stages: List[Operator] = Field(
-        required=True,
-        description="In-order list of operators that make up this pipeline",
-    )
-    router: Router = Field(
-        default_factor=Router,
-        description="Router object to determine order and run the stages. "
-        "Defaults to the base Router object",
-    )
-    schedulers: List[OperatorScheduler] = Field(
-        default_factor=lambda: [OperatorScheduler()],
-        description="List of schedulers to run operators in order of priority",
-    )
+    :param ops: Operators to run within the pipeline. Can either be a list of operators
+    or dictionary of operators.
+    :param router: A Router which dictates the next operator to call.
+    :param schedulers: A list of schedulers to run operators.
 
-    _scheduler_group: SchedulerGroup = PrivateAttr()
-
-    class Config:
-        arbitrary_types_allowed = True
+    """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(
+        self,
+        ops: Union[Dict[str, Operator], List[Operator]],
+        router: Router,
+        schedulers: List[OperatorScheduler],
+    ):
 
+        self.ops = ops
+        self.router = router
+        self.schedulers = schedulers
         self.validate()
 
         # SchedulerGroup handles running all schedulers in order of priority
         self._scheduler_group = SchedulerGroup(self.schedulers)
 
+    def run(self, inp: Any, context: Optional[Context]):
+        """
+        Run through the operators using the provided router and scheduler. Update the
+        context to reflect each step of the router. The input to a given operator is the
+        output of the previous operator.
+
+        :param inp: input to the operator. expected to be of any type that is
+        expected by the operator.
+        :param context: context to store the current the inputs, outputs, and operator
+        for each step of the router.
+
+        """
+        next_step = self.router.START_ROUTE
+        while next_step != self.router.END_ROUTE:
```

Comment:
Just to make sure: even though we linearly dispatch the consecutive router steps, in reality, the operators are not run sequentially - this all depends on the router's logic.

Reply:
not sure I follow your question

Reply:
I think I have a similar question, for instance in the image classification pipline we define the operators:
`ops = [preproces, engine, postprocess]` as a list, and could run the pipeline just stepping through the operators in list order. Iis the purpose of the router for cases where two operators can be run in parallel? `Does the order of ops matter at all?

Reply:
Ah ok I understand now looking through the router code, we have a base router that just runs the ops linearly. I think I'm still unclear about the use case for multiple types of routers though, since it looks like we are running operators sequentially and not allowing for computing them in parallel? In what case is the linear router not sufficient?

Reply:
I think the `GraphRouter` is the answer to my question :)